### PR TITLE
VRP Fixes and Tests

### PR DESCRIPTION
**Pull request by somzzz**
Added missing tests for https://github.com/dlang/dmd/pull/7355 and made a couple of adjustments. There were some issues with the sign of a number.

### CODE REVIEW

**Comment by ibuclaw** [comment]
Maybe this is one module that could benefit from having unittests to verify the internal implementation as well.

**Comment by JinShil** [comment]
I can't find any particular problem with this PR, but it is lacking an explanation describing what problem it solves, making it difficult to review.  Is there a bug in the current implementation, or is this just making the implementation more robust?

**Comment by somzzz** [comment]
@JinShil When https://github.com/dlang/dmd/pull/7355 was pulled, it was missing a couple of tests to ensure coverage for the new code. @andralex told me to add them in a different PR. That's the reason for this one.

File: src/dmd/intrange.d [hunk 1]
```diff
@@ -761,11 +770,26 @@ private:
...
         auto rhsc = IntRange(rhs);
 
+        // Sign bit not part of the .value so we need an extra iteration
+        if (lhsc.imax.negative ^ rhsc.imax.negative)
+        {
+            sign = true;
+            if (lhsc.imax.negative)
+            {
+                if (!lhsc.imin.negative) lhsc.imin.value = 0;
+                if (!rhsc.imin.negative) rhsc.imin.value = 0;
```
Comment (ibuclaw) [inline src/dmd/intrange.d]: Condition and branch on separate lines

File: src/dmd/intrange.d [hunk 2]
```diff
@@ -761,11 +770,26 @@ private:
...
+            }
+        }
+        else if (lhsc.imin.negative & rhsc.imin.negative) sign = true;
```
Comment (ibuclaw) [inline src/dmd/intrange.d]: Same here

File: src/dmd/intrange.d [hunk 3]
```diff
@@ -788,25 +812,30 @@ private:
...
     // https://github.com/tgehr/d-compiler/blob/master/vrange.d
     static SignExtendedNumber maxAnd(const IntRange lhs, const IntRange rhs)
     {
         uinteger_t x = 0;
+        bool sign = false;
         auto lhsc = IntRange(lhs);
         auto rhsc = IntRange(rhs);
 
-        for (uinteger_t d = 1LU << (8 * uinteger_t.sizeof-  1); d; d >>= 1)
+        if (lhsc.imax.negative & rhsc.imax.negative) sign = true;
```
Comment (ibuclaw) [inline src/dmd/intrange.d]: And here

File: src/dmd/intrange.d [hunk 4]
```diff
@@ -545,28 +548,32 @@ struct IntRange
...
 
-            return IntRange(minAnd(l, r), max);
+            auto range = IntRange(minAnd(l, r), max);
+            return range;
         }
         else
         {
             // only one interval spans [-1,0]
-            if ((l.imin.value ^ l.imax.value) < 0) swap(l, r); // r spans [-1,0]
+            if ((l.imin.negative ^ l.imax.negative) == 1) swap(l, r); // r spans [-1,0]
```
Comment (ibuclaw) [inline src/dmd/intrange.d]: And here

**Review by ibuclaw — CHANGES_REQUESTED** [review]
Needs rebasing

**Comment by andralex** [comment]
@somzzz can you please look into the autotester errors?

**Comment by somzzz** [comment]
@andralex when I click on ```Details``` it says build->passing. I don't see anything on the job logs.
@ibuclaw Do you have other comments about this PR?

**Comment by wilzbach** [comment]
Yeah auto-tester is full of random failures, see also: https://github.com/braddr/d-tester/issues/66

I closed and re-opened to let SemaphoreCI know about this PR. If the failing Travis is bothering you, simply rebase as we removed Travis completely.

**Comment by andralex** [comment]
I'll force merge this as the error seems unrelated. We should make it a top priority to have tests reliably pass or fail (hope these aren't the famous last words...).

**Pull request merged**